Source/NSURLSession.m

```diff
@@ -632,7 +681,7 @@ - (void) URLProtocol: (NSURLProtocol *)protocol
 
 @implementation NSURLSessionTask
 {
-  NSURLSession                   *_session; /* not retained */
+  NSURLSession                   *_session;
```

Comment:
Are you sure that you do not create a retain cycle by having the task retain the session? I don't know the rest of the code here but would expect that there was a reason why the session was treated special in the old code.

Reply:
The NSURLSession API is asynchronous. With ARC enabled the NSURLSession object gets released shortly after resume is called at the end of the method. I have to retain the session to keep up with the spec.

``` Objective-C
- (void) resultsForURL: (NSURL *) url 
{
  NSURLSessionConfiguration *defaultConfigObject;
  NSURLSessionDataTask      *dataTask;
  NSURLSession              *session;
 
  defaultConfigObject = [NSURLSessionConfiguration defaultSessionConfiguration];
  session = [NSURLSession sessionWithConfiguration: defaultConfigObject
                                          delegate: self
                                     delegateQueue: [NSOperationQueue mainQueue]];
  dataTask = [session dataTaskWithURL: url];
  
  [dataTask resume];
}
```

Reply:
I understand this issue, but this does not remove the problem you have created. Now the session retains its tasks and the task retains it session. Neither will ever be removed.
By the way, if you retain the session you also need to assign it in the copy method.

Reply:
Logically you'd expect the session to own its tasks (ie the task would not retain its session).  However, searching around I find confirmation that the session does seem to be retained in the Apple implementation, and I suppose it makes sense for convenience if you want to keep a reference to the task but not the session it belongs to.
However, we *do* need to avoid a retain loop causing session and tasks to be leaked.
I guess the way to do that might be to have the task only retain the session once the -resume method is called (ie the operation is started), and release the session as the last thing done once the task has completed.  So if a task is never started (-resume is never called) the session would not be retained, and there would be no retain loop, and similarly when the task stops executing the session would no longer be retained and the retain loop would be broken.